File: src/com/keepassdroid/database/PwDatabaseV4.java
/*
 * Copyright 2010-2013 Brian Pellin.
 *     
 * This file is part of KeePassDroid.
 *
 *  KeePassDroid is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  KeePassDroid is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePassDroid.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.keepassdroid.database;

import java.io.FileInputStream;
import java.io.IOException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.w3c.dom.Text;

import biz.source_code.base64Coder.Base64Coder;

import com.keepassdroid.database.exception.InvalidKeyFileException;


public class PwDatabaseV4 extends PwDatabase {

	public static final Date DEFAULT_NOW = new Date();
	private static final int DEFAULT_HISTORY_MAX_ITEMS = 10; // -1 unlimited
	private static final long DEFAULT_HISTORY_MAX_SIZE = 6 * 1024 * 1024; // -1 unlimited
	
	public UUID dataCipher;
	public PwCompressionAlgorithm compressionAlgorithm;
    public long numKeyEncRounds;
    public Date nameChanged;
    public String description;
    public Date descriptionChanged;
    public String defaultUserName;
    public Date defaultUserNameChanged;
    
    public Date keyLastChanged = DEFAULT_NOW;
    public long keyChangeRecDays = -1;
    public long keyChangeForceDays = 1;
    
    public long maintenanceHistoryDays = 365;
    public String color = "";
    public boolean recycleBinEnabled;
    public UUID recycleBinUUID;
    public Date recycleBinChanged;
    public UUID entryTemplatesGroup;
    public Date entryTemplatesGroupChanged;
    public int historyMaxItems = DEFAULT_HISTORY_MAX_ITEMS;
    public long historyMaxSize = DEFAULT_HISTORY_MAX_SIZE;
    public UUID lastSelectedGroup;
    public UUID lastTopVisibleGroup;
    public MemoryProtectionConfig memoryProtection = new MemoryProtectionConfig();
    public List<PwDeletedObject> deletedObjects = new ArrayList<PwDeletedObject>();
    public List<PwIconCustom> customIcons = new ArrayList<PwIconCustom>();
    public Map<String, String> customData = new HashMap<String, String>();
    public Map<String, byte[]> binPool = new HashMap<String, byte[]>();
    public Map<String, byte[]> binPoolCopyOnRead = new HashMap<String, byte[]>();
    
    public String localizedAppName = "KeePassDroid";
    
    public class MemoryProtectionConfig {
    	public boolean protectTitle = false;
    	public boolean protectUserName = false;
    	public boolean protectPassword = false;
    	public boolean protectUrl = false;
    	public boolean protectNotes = false;
    	
    	public boolean autoEnableVisualHiding = false;
    	
    	public boolean GetProtection(String field) {
    		if ( field.equalsIgnoreCase(PwDefsV4.TITLE_FIELD)) return protectTitle;
    		if ( field.equalsIgnoreCase(PwDefsV4.USERNAME_FIELD)) return protectUserName;
    		if ( field.equalsIgnoreCase(PwDefsV4.PASSWORD_FIELD)) return protectPassword;
    		if ( field.equalsIgnoreCase(PwDefsV4.URL_FIELD)) return protectUrl;
    		if ( field.equalsIgnoreCase(PwDefsV4.NOTES_FIELD)) return protectNotes;
    		
    		return false;
    	}
    }

    public static final UUID UUID_ZERO = new UUID(0,0);
    
	@Override
	public byte[] getMasterKey(String key, String keyFileName)
			throws InvalidKeyFileException, IOException {
		assert( key != null && keyFileName != null );
		
		byte[] fKey;
		
		if ( key.length() > 0 && keyFileName.length() > 0 ) {
			return getCompositeKey(key, keyFileName);
		} else if ( key.length() > 0 ) {
			fKey =  getPasswordKey(key);
		} else if ( keyFileName.length() > 0 ) {
			fKey = getFileKey(keyFileName);
		} else {
			throw new IllegalArgumentException( "Key cannot be empty." );
		}
		
		MessageDigest md;
		try {
			md = MessageDigest.getInstance("SHA-256");
		} catch (NoSuchAlgorithmException e) {
			throw new IOException("No SHA-256 implementation");
		}
		
		return md.digest(fKey);
	}

    @Override
	public byte[] getPasswordKey(String key) throws IOException {
		return getPasswordKey(key, "UTF-8");
	}
    
	private static final String RootElementName = "KeyFile";
	//private static final String MetaElementName = "Meta";
	//private static final String VersionElementName = "Version";
	private static final String KeyElementName = "Key";
	private static final String KeyDataElementName = "Data";
	
	@Override
	protected byte[] loadXmlKeyFile(String fileName) {
		try {
			DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
			DocumentBuilder db = dbf.newDocumentBuilder();
			FileInputStream fis = new FileInputStream(fileName);
			Document doc = db.parse(fis);
			
			Element el = doc.getDocumentElement();
			if (el == null || ! el.getNodeName().equalsIgnoreCase(RootElementName)) {
				return null;
			}
			
			NodeList children = el.getChildNodes();
			if (children.getLength() < 2) {
				return null;
			}
			
			for ( int i = 0; i < children.getLength(); i++ ) {
				Node child = children.item(i);
				
				if ( child.getNodeName().equalsIgnoreCase(KeyElementName) ) {
					NodeList keyChildren = child.getChildNodes();
					for ( int j = 0; j < keyChildren.getLength(); j++ ) {
						Node keyChild = keyChildren.item(j);
						if ( keyChild.getNodeName().equalsIgnoreCase(KeyDataElementName) ) {
							NodeList children2 = keyChild.getChildNodes();
							for ( int k = 0; k < children2.getLength(); k++) {
								Node text = children2.item(k);
								if (text.getNodeType() == Node.TEXT_NODE) {
									Text txt = (Text) text;
									return Base64Coder.decode(txt.getNodeValue());
								}
							}
						}
					}
				}
			}
		} catch (Exception e) {
			return null;
		}
		return null;
	}

	@Override
	public List<PwGroup> getGroups() {
		List<PwGroup> list = new ArrayList<PwGroup>();
		PwGroupV4 root = (PwGroupV4) rootGroup;
		root.buildChildGroupsRecursive(list);
		
		return list;
	}

	@Override
	public List<PwGroup> getGrpRoots() {
		return rootGroup.childGroups;
	}

	@Override
	public List<PwEntry> getEntries() {
		List<PwEntry> list = new ArrayList<PwEntry>();
		PwGroupV4 root = (PwGroupV4) rootGroup;
		root.buildChildEntriesRecursive(list);
		
		return list;
	}

	@Override
	public long getNumRounds() {
		return numKeyEncRounds;
	}

	@Override
	public void setNumRounds(long rounds) throws NumberFormatException {
		numKeyEncRounds = rounds;
		
	}

	@Override
	public boolean appSettingsEnabled() {
		return false;
	}

	@Override
	public PwEncryptionAlgorithm getEncAlgorithm() {
		return PwEncryptionAlgorithm.Rjindal;
	}

	@Override
	public PwGroupIdV4 newGroupId() {
		PwGroupIdV4 id = new PwGroupIdV4(UUID_ZERO);
		
		while (true) {
			id = new PwGroupIdV4(UUID.randomUUID());
			
			if (!isGroupIdUsed(id)) break;
		}
		
		return id;
	}

	@Override
	public PwGroup createGroup() {
		return new PwGroupV4();
	}

	@Override
	public boolean isBackup(PwGroup group) {
		if (!recycleBinEnabled) {
			return false;
		}
		PwGroupV4 g = (PwGroupV4) group;
		
		// Need to loop upwards to see if any ancestor is the recycle bin
		while (g != null) {
			if (recycleBinUUID.equals(g.uuid) || g.name.equalsIgnoreCase("Backup")) {
				return true;
			}
			
			g = g.parent;
		}
		
		return false;
	}

	@Override
	public void populateGlobals(PwGroup currentGroup) {
		groups.put(rootGroup.getId(), rootGroup);
		
		super.populateGlobals(currentGroup);
	}
}	

File: app/src/main/java/com/kunzisoft/keepass/database/element/PwDatabaseV4.kt
/*
 * Copyright 2019 Jeremy Jamet / Kunzisoft.
 *     
 * This file is part of KeePass DX.
 *
 *  KeePass DX is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  KeePass DX is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with KeePass DX.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.kunzisoft.keepass.database.element

import android.content.res.Resources
import android.util.Log
import biz.source_code.base64Coder.Base64Coder
import com.kunzisoft.keepass.R
import com.kunzisoft.keepass.crypto.CryptoUtil
import com.kunzisoft.keepass.crypto.engine.AesEngine
import com.kunzisoft.keepass.crypto.engine.CipherEngine
import com.kunzisoft.keepass.crypto.keyDerivation.*
import com.kunzisoft.keepass.database.exception.InvalidKeyFileException
import com.kunzisoft.keepass.database.exception.UnknownKDF
import com.kunzisoft.keepass.database.file.PwCompressionAlgorithm
import com.kunzisoft.keepass.utils.VariantDictionary
import org.w3c.dom.Node
import org.w3c.dom.Text
import java.io.IOException
import java.io.InputStream
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.util.*
import javax.xml.parsers.DocumentBuilderFactory


class PwDatabaseV4 : PwDatabase<PwGroupV4, PwEntryV4> {

    var hmacKey: ByteArray? = null
        private set
    var dataCipher = AesEngine.CIPHER_UUID
    private var dataEngine: CipherEngine = AesEngine()
    var compressionAlgorithm = PwCompressionAlgorithm.Gzip
    var kdfParameters: KdfParameters? = null
    private var kdfV4List: MutableList<KdfEngine> = ArrayList()
    private var numKeyEncRounds: Long = 0
    var publicCustomData = VariantDictionary()

    var name = "KeePass DX database"
    var nameChanged = PwDate()
    // TODO change setting date
    var settingsChanged = PwDate()
    var description = ""
    var descriptionChanged = PwDate()
    var defaultUserName = ""
    var defaultUserNameChanged = PwDate()

    // TODO date
    var keyLastChanged = PwDate()
    var keyChangeRecDays: Long = -1
    var keyChangeForceDays: Long = 1
    var isKeyChangeForceOnce = false

    var maintenanceHistoryDays: Long = 365
    var color = ""
    /**
     * Determine if RecycleBin is enable or not
     * @return true if RecycleBin enable, false if is not available or not enable
     */
    var isRecycleBinEnabled = true
    var recycleBinUUID: UUID = UUID_ZERO
    var recycleBinChanged = Date()
    var entryTemplatesGroup = UUID_ZERO
    var entryTemplatesGroupChanged = PwDate()
    var historyMaxItems = DEFAULT_HISTORY_MAX_ITEMS
    var historyMaxSize = DEFAULT_HISTORY_MAX_SIZE
    var lastSelectedGroupUUID = UUID_ZERO
    var lastTopVisibleGroupUUID = UUID_ZERO
    var memoryProtection = MemoryProtectionConfig()
    val deletedObjects = ArrayList<PwDeletedObject>()
    val customIcons = ArrayList<PwIconCustom>()
    val customData = HashMap<String, String>()

    var binPool = BinaryPool()

    var localizedAppName = "KeePassDX" // TODO resource

    init {
        kdfV4List.add(KdfFactory.aesKdf)
        kdfV4List.add(KdfFactory.argon2Kdf)
    }

    constructor()

    constructor(databaseName: String) {
        val groupV4 = createGroup().apply {
            title = databaseName
            icon = iconFactory.folderIcon
        }
        rootGroup = groupV4
        addGroupIndex(groupV4)
    }

    override val version: String
        get() = "KeePass 2"

    override val kdfEngine: KdfEngine?
        get() = try {
            getEngineV4(kdfParameters)
        } catch (unknownKDF: UnknownKDF) {
            Log.i(TAG, "Unable to retrieve KDF engine", unknownKDF)
            null
        }

    override val kdfAvailableList: List<KdfEngine>
        get() = kdfV4List

    @Throws(UnknownKDF::class)
    fun getEngineV4(kdfParameters: KdfParameters?): KdfEngine {
        val unknownKDFException = UnknownKDF()
        if (kdfParameters == null) {
            throw unknownKDFException
        }
        for (engine in kdfV4List) {
            if (engine.uuid == kdfParameters.uuid) {
                return engine
            }
        }
        throw unknownKDFException
    }

    override val availableEncryptionAlgorithms: List<PwEncryptionAlgorithm>
        get() {
            val list = ArrayList<PwEncryptionAlgorithm>()
            list.add(PwEncryptionAlgorithm.AESRijndael)
            list.add(PwEncryptionAlgorithm.Twofish)
            list.add(PwEncryptionAlgorithm.ChaCha20)
            return list
        }

    override var numberKeyEncryptionRounds: Long
        get() {
            val kdfEngine = kdfEngine
            if (kdfEngine != null && kdfParameters != null)
                numKeyEncRounds = kdfEngine.getKeyRounds(kdfParameters!!)
            return numKeyEncRounds
        }
        @Throws(NumberFormatException::class)
        set(rounds) {
            val kdfEngine = kdfEngine
            if (kdfEngine != null && kdfParameters != null)
                kdfEngine.setKeyRounds(kdfParameters!!, rounds)
            numKeyEncRounds = rounds
        }

    var memoryUsage: Long
        get() {
            val kdfEngine = kdfEngine
            return if (kdfEngine != null && kdfParameters != null) {
                kdfEngine.getMemoryUsage(kdfParameters!!)
            } else KdfEngine.UNKNOWN_VALUE.toLong()
        }
        set(memory) {
            val kdfEngine = kdfEngine
            if (kdfEngine != null && kdfParameters != null)
                kdfEngine.setMemoryUsage(kdfParameters!!, memory)
        }

    var parallelism: Int
        get() {
            val kdfEngine = kdfEngine
            return if (kdfEngine != null && kdfParameters != null) {
                kdfEngine.getParallelism(kdfParameters!!)
            } else KdfEngine.UNKNOWN_VALUE
        }
        set(parallelism) {
            val kdfEngine = kdfEngine
            if (kdfEngine != null && kdfParameters != null)
                kdfEngine.setParallelism(kdfParameters!!, parallelism)
        }

    override val passwordEncoding: String
        get() = "UTF-8"

    fun getGroupByUUID(groupUUID: UUID): PwGroupV4? {
        if (groupUUID == UUID_ZERO)
            return null
        return getGroupById(PwNodeIdUUID(groupUUID))
    }

    // Retrieve recycle bin in index
    val recycleBin: PwGroupV4?
        get() = getGroupByUUID(recycleBinUUID)

    val lastSelectedGroup: PwGroupV4?
        get() = getGroupByUUID(lastSelectedGroupUUID)

    val lastTopVisibleGroup: PwGroupV4?
        get() = getGroupByUUID(lastTopVisibleGroupUUID)

    fun setDataEngine(dataEngine: CipherEngine) {
        this.dataEngine = dataEngine
    }

    fun getCustomIcons(): List<PwIconCustom> {
        return customIcons
    }

    fun addCustomIcon(customIcon: PwIconCustom) {
        this.customIcons.add(customIcon)
    }

    fun getCustomData(): Map<String, String> {
        return customData
    }

    fun putCustomData(label: String, value: String) {
        this.customData[label] = value
    }

    override fun containsCustomData(): Boolean {
        return getCustomData().isNotEmpty()
    }

    @Throws(InvalidKeyFileException::class, IOException::class)
    public override fun getMasterKey(key: String?, keyInputStream: InputStream?): ByteArray {

        var masterKey = byteArrayOf()

        if (key != null && keyInputStream != null) {
            return getCompositeKey(key, keyInputStream)
        } else if (key != null) { // key.length() >= 0
            masterKey = getPasswordKey(key)
        } else if (keyInputStream != null) { // key == null
            masterKey = getFileKey(keyInputStream)
        }

        val messageDigest: MessageDigest
        try {
            messageDigest = MessageDigest.getInstance("SHA-256")
        } catch (e: NoSuchAlgorithmException) {
            throw IOException("No SHA-256 implementation")
        }

        return messageDigest.digest(masterKey)
    }

    @Throws(IOException::class)
    fun makeFinalKey(masterSeed: ByteArray) {

        kdfParameters?.let { keyDerivationFunctionParameters ->
            val kdfEngine = getEngineV4(keyDerivationFunctionParameters)

            var transformedMasterKey = kdfEngine.transform(masterKey, keyDerivationFunctionParameters)
            if (transformedMasterKey.size != 32) {
                transformedMasterKey = CryptoUtil.hashSha256(transformedMasterKey)
            }

            val cmpKey = ByteArray(65)
            System.arraycopy(masterSeed, 0, cmpKey, 0, 32)
            System.arraycopy(transformedMasterKey, 0, cmpKey, 32, 32)
            finalKey = CryptoUtil.resizeKey(cmpKey, 0, 64, dataEngine.keyLength())

            val messageDigest: MessageDigest
            try {
                messageDigest = MessageDigest.getInstance("SHA-512")
                cmpKey[64] = 1
                hmacKey = messageDigest.digest(cmpKey)
            } catch (e: NoSuchAlgorithmException) {
                throw IOException("No SHA-512 implementation")
            } finally {
                Arrays.fill(cmpKey, 0.toByte())
            }
        }
    }

    override fun loadXmlKeyFile(keyInputStream: InputStream): ByteArray? {
        try {
            val dbf = DocumentBuilderFactory.newInstance()
            val db = dbf.newDocumentBuilder()
            val doc = db.parse(keyInputStream)

            val el = doc.documentElement
            if (el == null || !el.nodeName.equals(RootElementName, ignoreCase = true)) {
                return null
            }

            val children = el.childNodes
            if (children.length < 2) {
                return null
            }

            for (i in 0 until children.length) {
                val child = children.item(i)

                if (child.nodeName.equals(KeyElementName, ignoreCase = true)) {
                    val keyChildren = child.childNodes
                    for (j in 0 until keyChildren.length) {
                        val keyChild = keyChildren.item(j)
                        if (keyChild.nodeName.equals(KeyDataElementName, ignoreCase = true)) {
                            val children2 = keyChild.childNodes
                            for (k in 0 until children2.length) {
                                val text = children2.item(k)
                                if (text.nodeType == Node.TEXT_NODE) {
                                    val txt = text as Text
                                    return Base64Coder.decode(txt.nodeValue)
                                }
                            }
                        }
                    }
                }
            }
        } catch (e: Exception) {
            return null
        }

        return null
    }

    override fun newGroupId(): PwNodeIdUUID {
        var newId: PwNodeIdUUID
        do {
            newId = PwNodeIdUUID()
        } while (isGroupIdUsed(newId))

        return newId
    }

    override fun newEntryId(): PwNodeIdUUID {
        var newId: PwNodeIdUUID
        do {
            newId = PwNodeIdUUID()
        } while (isEntryIdUsed(newId))

        return newId
    }

    override fun createGroup(): PwGroupV4 {
        return PwGroupV4()
    }

    override fun createEntry(): PwEntryV4 {
        return PwEntryV4()
    }

    override fun rootCanContainsEntry(): Boolean {
        return true
    }

    override fun isBackup(group: PwGroupV4): Boolean {
        // To keep compatibility with old V1 databases
        var currentGroup: PwGroupV4? = group
        while (currentGroup != null) {
            if (currentGroup.parent == rootGroup
                    && currentGroup.title.equals("Backup", ignoreCase = true)) {
                return true
            }
            currentGroup = currentGroup.parent
        }

        return if (recycleBin == null)
            false
        else if (!isRecycleBinEnabled)
            false
        else
            group.isContainedIn(recycleBin!!)
    }

    /**
     * Ensure that the recycle bin tree exists, if enabled and create it
     * if it doesn't exist
     */
    private fun ensureRecycleBin(resources: Resources) {
        if (recycleBin == null) {
            // Create recycle bin
            val recycleBinGroup = createGroup().apply {
                title = resources.getString(R.string.recycle_bin)
                icon = iconFactory.trashIcon
                enableAutoType = false
                enableSearching = false
                isExpanded = false
            }
            addGroupTo(recycleBinGroup, rootGroup)
            recycleBinUUID = recycleBinGroup.id
            recycleBinChanged = recycleBinGroup.lastModificationTime.date
        }
    }

    /**
     * Define if a Node must be delete or recycle when remove action is called
     * @param node Node to remove
     * @return true if node can be recycle, false elsewhere
     */
    fun canRecycle(node: PwNode<*, PwGroupV4, PwEntryV4>): Boolean {
        if (!isRecycleBinEnabled)
            return false
        if (recycleBin == null)
            return true
        if (!node.isContainedIn(recycleBin!!))
            return true
        return false
    }

    fun recycle(group: PwGroupV4, resources: Resources) {
        ensureRecycleBin(resources)
        removeGroupFrom(group, group.parent)
        addGroupTo(group, recycleBin)
        group.afterAssignNewParent()
    }

    fun recycle(entry: PwEntryV4, resources: Resources) {
        ensureRecycleBin(resources)
        removeEntryFrom(entry, entry.parent)
        addEntryTo(entry, recycleBin)
        entry.afterAssignNewParent()
    }

    fun undoRecycle(group: PwGroupV4, origParent: PwGroupV4) {
        removeGroupFrom(group, recycleBin)
        addGroupTo(group, origParent)
    }

    fun undoRecycle(entry: PwEntryV4, origParent: PwGroupV4) {
        removeEntryFrom(entry, recycleBin)
        addEntryTo(entry, origParent)
    }

    fun getDeletedObjects(): List<PwDeletedObject> {
        return deletedObjects
    }

    fun addDeletedObject(deletedObject: PwDeletedObject) {
        this.deletedObjects.add(deletedObject)
    }

    override fun removeEntryFrom(entryToRemove: PwEntryV4, parent: PwGroupV4?) {
        super.removeEntryFrom(entryToRemove, parent)
        deletedObjects.add(PwDeletedObject(entryToRemove.id))
    }

    override fun undoDeleteEntryFrom(entry: PwEntryV4, origParent: PwGroupV4?) {
        super.undoDeleteEntryFrom(entry, origParent)
        deletedObjects.remove(PwDeletedObject(entry.id))
    }

    fun containsPublicCustomData(): Boolean {
        return publicCustomData.size() > 0
    }

    override fun validatePasswordEncoding(key: String?): Boolean {
        if (key == null)
            return true
        return super.validatePasswordEncoding(key)
    }

    override fun clearCache() {
        super.clearCache()
        binPool.clear()
    }

    companion object {
        private val TAG = PwDatabaseV4::class.java.name

        private const val DEFAULT_HISTORY_MAX_ITEMS = 10 // -1 unlimited
        private const val DEFAULT_HISTORY_MAX_SIZE = (6 * 1024 * 1024).toLong() // -1 unlimited

        private const val RootElementName = "KeyFile"
        //private const val MetaElementName = "Meta";
        //private const val VersionElementName = "Version";
        private const val KeyElementName = "Key"
        private const val KeyDataElementName = "Data"
    }
}